File: materials/forecast/python/controllers/weather.py
from decimal import Decimal
from typing import Tuple, Optional
from clients.weather_data_client import WeatherDataClient
from models.weather.get import CurrentWeather

class CurrentWeatherController:
    def __init__(self, client: WeatherDataClient):
        self.client = client
    
    def get_current_weather(self, lat: Decimal, lon: Decimal) -> Tuple[CurrentWeather, Optional[Exception]]:
        temperature, err = self.client.location_current_temperature(lat, lon)
        if err:
            return CurrentWeather(Decimal('0')), err
        
        return CurrentWeather(temperature), None

File: materials/forecast/rust/controllers/src/weather.rs
use std::error::Error;

use clients::WeatherDataClient;
use models::weather::CurrentWeather;
use rust_decimal::Decimal;

#[derive(Clone)]
pub struct CurrentWeatherController<T: WeatherDataClient + Clone> {
    pub client: T,
}

impl<T: WeatherDataClient + Clone> CurrentWeatherController<T> {
    pub fn new(client: T) -> Self {
        CurrentWeatherController { client }
    }
    pub async fn get_current_weather(
        &self,
        lat: Decimal,
        lon: Decimal,
    ) -> Result<CurrentWeather, Box<dyn Error>> {
        Ok(CurrentWeather {
            temperature: self.client.location_current_temperature(lat, lon).await?,
        })
    }
}
